canvas_oauth/tests/test_oauth.py

```diff
@@ -0,0 +1,240 @@
+        query_params = {
+            "q": "education",
+            "page": "5"
+        }
+        request = rf.get('/index', data=query_params)
+        request.user = MagicMock()
+        request.session = {}
+
+        redirect_uri = request.build_absolute_uri(reverse('canvas-oauth-callback'))
+        login_url = get_oauth_login_url(
+            client_id=settings.CANVAS_OAUTH_CLIENT_ID,
+            redirect_uri=redirect_uri,
+            state=request_state)
+
+        # run tests
+        response = handle_missing_token(request)
+
+        # check the response
+        self.assertEqual(302, response.status_code)
+        self.assertEqual(login_url, response['Location'])
+
+        # check the session values
+        self.assertEqual(request.get_full_path(), request.session['canvas_oauth_initial_uri'])
+        self.assertEqual(request_state, request.session['canvas_oauth_request_state'])
+        self.assertEqual(redirect_uri, request.session["canvas_oauth_redirect_uri"])
+
+
+class TestGetOauthToken(TestCase):
+
+    def get_mock_request_with_token(self, expired=False, **kwargs):
+        access_token = kwargs.get("access_token", "access-token-123")
+        refresh_token = kwargs.get("refresh_token", "refresh-token-abc")
+        expires = kwargs.get("expires", datetime(2001, 1, 1) if expired else timezone.now() + timedelta(seconds=100))
+
+        stub_canvas_oauth2_token = StubCanvasOAuth2Token(access_token, refresh_token, expires)
+        stub_canvas_oauth2_token.stub_expires_within_return_value(expired)
+
+        mock_user = MagicMock()
+        type(mock_user).canvas_oauth2_token = PropertyMock(return_value=stub_canvas_oauth2_token)
+
+        request = RequestFactory().get('/index')
+        request.user = mock_user
+
+        return request
+
+    @patch('canvas_oauth.oauth.refresh_oauth_token')
+    def test_unexpired_access_token(self, mock_refresh_oauth_token):
+        expected_access_token = "unexpired-access-token-123"
+        request = self.get_mock_request_with_token(expired=False, access_token=expected_access_token)
+        actual_access_token = get_oauth_token(request)
+        self.assertFalse(mock_refresh_oauth_token.called)
+        self.assertEqual(expected_access_token, actual_access_token)
+
+    @patch('canvas_oauth.oauth.refresh_oauth_token')
+    def test_expired_access_token(self, mock_refresh_oauth_token):
+        request = self.get_mock_request_with_token(expired=True)
+        get_oauth_token(request)
+        mock_refresh_oauth_token.assert_called_with(request)
+
+        stub_canvas_oauth2_token = request.user.canvas_oauth2_token
+        expires_buffer = settings.CANVAS_OAUTH_TOKEN_EXPIRATION_BUFFER
+        self.assertTrue(stub_canvas_oauth2_token.stub_expires_within_called_with(expires_buffer))
+
+    def test_missing_token_error(self):
+        mock_user = MagicMock()
+        type(mock_user).canvas_oauth2_token = PropertyMock(side_effect=CanvasOAuth2Token.DoesNotExist())
```

Comment:
😮 It might be nice for a future `mock`er to have a description of this magic with `type` and `PropertyMock`? It may be a common idiom, but I can't parse it at first glance. (It's used a few times in these tests.)

Reply:
Good point, it's a bit of an odd idiom that's new to me, and maybe too obscure to be worth using. In any event, I'll add a comment to that affect for future code readers, taken from the unittest documentation: 
 https://docs.python.org/3/library/unittest.mock.html#unittest.mock.PropertyMock

Reply:
Yeah I think in part it's the `type()` that's making my head hurt here. Good to see it's directly in the `unittest` docs.